KAFKA-2071: Replace Producer Request/Response with their org.apache.kafka.common.requests equivalents

dajac opened the pull request:
This PR replaces all occurrences of kafka.api.ProducerRequest/ProducerResponse by their common equivalents.

dajac committed:
Replace Producer Request/Response with their org.apache.kafka.common.requests equivalents.

dajac commented:
@ijuma Ok, thanks. I'll update that.

dajac commented:
I've rebased it and removed implicit conversion of collections.

dajac commented:
Merged trunk and fixed conflicts.

dajac commented:
Failure is not related:

```
kafka.consumer.MetricsTest > testMetricsLeak FAILED
```

dajac commented:
Could someone review this one? May be @gwenshap or @ijuma?

granthenke commented:
@dajac Are you interested in rebasing this?

dajac commented:
@granthenke Yes, I'll have some time next week to do it.

dajac commented:
@granthenke I have rebased this PR. Note that this PR migrates the server and the producer.

dajac commented:
Migration of the producer is quite straightforward as it mainly happens in `SyncProducer` and `DefaultEventHandler` but I'm wondering if `SyncProducer` is considered as a public API. Is it?

granthenke commented:
@dajac It looks like the client code changes are not backward compatible. I think this patch either needs to be server side only, or if the clients are going to be updated, they need to be done so in a backward compatible way.

I could be wrong, as we have not had the discussion yet on the mailing list, but I think the idea is to deprecate the older Scala producers and consumers. I am not sure if its worth the effort of updating the client code. (context can be found in the comments on [KAFKA-2982](https://issues.apache.org/jira/browse/KAFKA-2982))

ijuma commented:
Looking at the JIRA for this PR, it looks like @gwenshap suggested that the JIRA was about removing k.a.Producer{Request, Response} altogether and that is why @dajac did this. But the producer is indeed public API and it is not yet deprecated so I think we can't do that for the upcoming release.

My suggestion is to only update the server-side and leave the soon to be deprecated clients alone (as @granthenke has suggested too, if I understand correctly).

dajac commented:
@granthenke @ijuma Indeed, the `SyncProducer` part is not backward compatible. Making it backward compatible requires to keep the old produce request so it doesn't make sense to upgrade it to the new one. As you both suggested, it is better to not upgrade it.

I'll update the PR to reflect this.

dajac commented:
@granthenke @ijuma I have rebased and updated it. Could you review it when you have time? All tests pass locally.

granthenke commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -315,44 +315,44 @@ class KafkaApis(val requestChannel: RequestChannel,
+            request.header.clientId,
+            topicPartition,
+            Errors.forCode(status.errorCode).exception.getClass.getName))
+          errorInResponse = true
```
Think this is likely just a rebase thing, but recently this was moved above the debug message. Not sure why, but can we keep it that way?

Actually just noticed its there, twice. We should remove this one.

dajac replied:
Indeed, I have forgotten to remove it during the rebase. I'll remove it.

ijuma replied:
I moved it above the debug log in the other PR to make the side-effect more prominent (instead of hidden after a long debug statement).

granthenke commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -315,44 +315,44 @@ class KafkaApis(val requestChannel: RequestChannel,
           if (errorInResponse) {
             val exceptionsSummary = mergedResponseStatus.map { case (topicAndPartition, status) =>
-              topicAndPartition -> Errors.forCode(status.error).exception.getClass.getName
+              topicAndPartition -> Errors.forCode(status.errorCode).exception.getClass.getName
```
Going to send a PR for this today or tomorrow (KAFKA-3019). Just an fyi.

dajac replied:
Ok, thanks.

granthenke commented on core/src/test/scala/unit/kafka/api/RequestResponseSerializationTest.scala:
```diff
@@ -65,7 +65,7 @@ object SerializationTestUtils {
       {
         case(partitionDataMessage, partition) =>
-          (TopicAndPartition(topic, partition), partitionDataMessage)
+          (new TopicAndPartition(topic, partition), partitionDataMessage)
```
Is this needed?

dajac replied:
No, it is not. I'll remove it.

granthenke commented:
Thanks for the update. I did a quick visual check and will follow up with a more in depth review likely tomorrow.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -122,6 +122,10 @@ public int timeout() {
         return partitionRecords;
     }
 
+    public void emptyData() {
```
I know this name was used in the old `ProduceRequest`, but it seems a bit ambiguous to me. Something with `clear` in the name would be clearer for me (`clearData`, `clearRecords`, `clearPartitionRecords`, etc.).

dajac replied:
Good point. I'll use `clearPartitionRecords`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -315,44 +315,44 @@ class KafkaApis(val requestChannel: RequestChannel,
-    val produceRequest = request.requestObj.asInstanceOf[ProducerRequest]
-    val numBytesAppended = produceRequest.sizeInBytes
+    val produceRequest = request.body.asInstanceOf[ProduceRequest]
+    val numBytesAppended = request.header.sizeOf() + produceRequest.sizeOf()
```
Nitpick: You don't need `()` in `sizeOf`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -360,35 +360,38 @@ class KafkaApis(val requestChannel: RequestChannel,
-          requestChannel.sendResponse(new RequestChannel.Response(request,
-                                                                  new RequestOrResponseSend(request.connectionId,
-                                                                                            response)))
+          val respHeader = new ResponseHeader(request.header.correlationId())
```
Nitpick: correlationId doesn't need `()`. Can you please check other similar cases too? (apiVersion and maybe others).

dajac replied:
Sure, I'll check them.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -360,35 +360,38 @@ class KafkaApis(val requestChannel: RequestChannel,
+      val internalTopicsAllowed = request.header.clientId == AdminUtils.AdminClientId
+
+      // Convert ByteBuffer to ByteBufferMessageSet
+      val authorizedMessagesPerPartition = authorizedRequestInfo.mapValues(new ByteBufferMessageSet(_))
```
`mapValues` is lazy (ie `ByteBufferMessageSet` will be created each time a Map value is accessed), is that what we want here?

dajac replied:
Oh. I was not aware of that. I think that it is better to use non-lazy way here.

ijuma commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -64,6 +66,13 @@ class SocketServerTest extends JUnitSuite {
     outgoing.flush()
   }
 
+  def sendRequest(socket: Socket, request: Array[Byte]) {
```
Would it not make sense to change the existing `sendRequest` so that it looks something like the following?

``` scala
  def sendRequest(socket: Socket, request: Array[Byte], id: Option[Short] = None) {
    val outgoing = new DataOutputStream(socket.getOutputStream)
    id match {
      case Some(id) =>
        outgoing.writeInt(request.length + 2)
        outgoing.writeShort(id)
      case None =>
        outgoing.writeInt(request.length)
    }
    outgoing.write(request)
    outgoing.flush()
  }
```

Alternatively, if we expect the other one to be removed once we transition to new request types, we should change it's name and add a comment to make that clear.

dajac replied:
I like your idea as both ways will stay. Thanks!

ijuma commented:
I had a look and left some comments. Looks good generally.

dajac committed:
Address reviews.

dajac commented:
@granthenke @ijuma I did a first pass and addressed all your reviews. Thanks for you input.

granthenke commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -190,30 +191,30 @@ class GroupMetadataManager(val brokerId: Int,
+      if (status.errorCode != Errors.NONE.code) {
         debug("Metadata from group %s with generation %d failed when appending to log due to %s"
-          .format(group.groupId, generationId, Errors.forCode(status.error).exception.getClass.getName))
+          .format(group.groupId, generationId, Errors.forCode(status.errorCode).exception.getClass.getName))
```
Can use exceptionName method from #754

granthenke commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -190,30 +191,30 @@ class GroupMetadataManager(val brokerId: Int,
 
           error("Appending metadata message for group %s generation %d failed due to %s, returning UNKNOWN error code to the client"
-            .format(group.groupId, generationId, Errors.forCode(status.error).exception.getClass.getName))
+            .format(group.groupId, generationId, Errors.forCode(status.errorCode).exception.getClass.getName))
```
Same as above

granthenke commented:
@dajac Just a minor update for the exceptionName method. Otherwise LGTM.

dajac commented:
@granthenke I have rebased the PR and addressed your points.

dajac commented:
@gwenshap @ewencp @junrao Could one of you take a look? @granthenke and @ijuma have reviewed.

ewencp commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -39,10 +38,11 @@ object RequestChannel extends Logging {
-    emptyProducerRequest.writeTo(byteBuffer)
+    val emptyRequestHeader = new RequestHeader(ApiKeys.PRODUCE.id, "", 0)
+    val emptyProduceRequest = new ProduceRequest(0, 0, new HashMap[TopicPartition, ByteBuffer]())
+    val byteBuffer = ByteBuffer.allocate(emptyRequestHeader.sizeOf() + emptyProduceRequest.sizeOf())
```
Not critical, but could we reuse `RequestSend.serialize()` here and avoid duplicating low-level serialization code?

dajac replied:
Yes, it is possible. I'll do it.

ewencp commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -360,41 +359,46 @@ class KafkaApis(val requestChannel: RequestChannel,
-                                                                  new RequestOrResponseSend(request.connectionId,
-                                                                                            response)))
+          val respHeader = new ResponseHeader(request.header.correlationId)
+          val respBody = if (request.header.apiVersion > 0)
```
Should we be more explicit here wrt versioning? As is, when we add new versions we can accidentally get the wrong behavior if we miss updating this piece of code since the `request.header.apiVersion > 0` will catch all new versions. Explicitly handling each version and generating an error for the unexpected case seems safer (which should be impossible for now since it would be caught at parse time, but would catch the case where a new version is added but this code is not updated). Presumably this code will always need some updates with modifications to this request/response pair, but better safe than sorry.

dajac replied:
Indeed, explicitly handling each version is safer. I'll update the logic and log an error/throw an exception when `apiVersion` is unknown.

ewencp commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -21,24 +21,26 @@ package kafka.network;
 import java.net._
 import javax.net.ssl._
 import java.io._
+import java.util
```
Doesn't look like this import is needed?

ewencp replied:
Sorry, rather, we should import `java.util.HashMap` rather than importing `java.util` and referring to it as `util.HashMap`.

dajac replied:
Good point.

dajac committed:
Address reviews.

dajac commented:
@ewencp I have addressed your points. Now, version is explicitly handled and an exception is thrown if the version is not supported. I'm just wondering if throwing an exception is the best way especially when `sendResponseCallback` is called from the purgatory. Also, I have updated `ProduceRequest.getErrorResponse` to handle version 0 and 1.

dajac committed:
Fix RequestResponseTest

dajac commented:
I have rebased and addressed all the comments.

ewencp commented:
LGTM, thanks @dajac!
